File: C++/BinaryTreePreorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
       vector<int> preOrder;
       preorderTraversalUtill(root, preOrder);
       return preOrder;
    }
    void preorderTraversalUtill(TreeNode* root, vector<int>& preOrder){
        if(root == NULL)
            return;
        preOrder.push_back(root->val);
        preorderTraversalUtill(root->left, preOrder);
        preorderTraversalUtill(root->right, preOrder);
        
    }
};

File: Java/BinaryTreePreorderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        ArrayList<Integer> preorder = new ArrayList<Integer>();
        
        if(root == null){
            return preorder;
        }
        
        Stack<TreeNode> stack = new Stack<TreeNode>();
        stack.push(root);
        
        TreeNode curr = null;
        TreeNode prev = null;
        
        while(!stack.isEmpty()){
            curr = stack.pop();
            
            if(prev == null || prev.left == curr || prev.right == curr){
                preorder.add(curr.val);
                
                if(curr.right != null){
                    stack.push(curr.right);
                }
                if(curr.left != null){
                    stack.push(curr.left);
                }
            }
            else{
                preorder.add(curr.val);
                
                if(curr.right != null){
                    stack.push(curr.right);
                }
                if(curr.left != null){
                    stack.push(curr.left);
                }
            }
            prev = curr;
            
        }
        return preorder;
    }
}
